File: Assets/src/model/tasks/WaitTask.cs
using System;
using System.Collections.Generic;
using UnityEngine;

/// <summary>Close-ended.</summary>
[System.Serializable]
public class WaitTask : ActorTask {
  public override TaskStage WhenToCheckIsDone => TaskStage.After;
  private int turns;

  public int Turns => turns;

  public WaitTask(Actor actor, int turns) : base(actor) {
    this.turns = turns;
  }

  protected override BaseAction GetNextActionImpl() {
    turns--;
    return new WaitBaseAction(actor);
  }

  public override bool IsDone() => turns <= 0;
}

File: web/src/model/tasks/WaitTask.ts
import { ActorTask, TaskStage } from '../ActorTask';
import { WaitBaseAction, type BaseAction } from '../BaseAction';
import type { Actor } from '../Actor';

export class WaitTask extends ActorTask {
  private turns: number;

  get whenToCheckIsDone(): TaskStage {
    return TaskStage.After;
  }

  get turnsRemaining(): number {
    return this.turns;
  }

  constructor(actor: Actor, turns: number) {
    super(actor);
    this.turns = turns;
  }

  protected getNextActionImpl(): BaseAction {
    this.turns--;
    return new WaitBaseAction(this.actor);
  }

  isDone(): boolean {
    return this.turns <= 0;
  }
}
